misc/clock.ru

```diff
@@ -41,30 +55,21 @@ class FlyHandler
   # Normal HTTP request handler.
   def call(env)
     # This part is only needed for Iodine.
-    if env['HTTP_UPGRADE'.freeze] =~ /websocket/i
+    if env['upgrade.websocket?'.freeze]
       env['upgrade.websocket'.freeze] = TickTock.new(env)
       return [0, {}, []]
     end
     [ 200, { }, [ "flying fish clock" ] ]
   end
 
-  # This method is called when a WebSockets or SSE connection is
-  # established. The env argument is the same as what would be given to the
-  # call method except the env[`push.writer'] is set to a writer that can be
-  # used to push or publish data to the browser.
-  #
-  # Return a handler for on_message and on_close callbacks. This is
-  # optional. If nil is returned then broadcasts to the browsers are possible
-  # but not direct write from Ruby are. If the call raises an exception then
-  # the connection is closed.
-  def on_open(env)
-    TickTock.new(env)
-  end
```

Comment:
This could be renamed but the approach I prefer to use for establishing a connection would call this instead of overloading the `call` method.

Reply:
Maybe I misunderstood - the `call` function is part of the Rack specification. We call the function once (as we must) before the upgrade, as the request comes in.

If we don't call the function for the HTTP upgrade request, we loose all the middleware and introduce Rack incompatibilities.

Consider authentication middleware used to authenticate the client using HTTP headers or cookies... this authentication should be performed before accepting Websocket / SSE connections. If we call `on_open(env`, we loose that.

Reply:
Some good points. What I was trying to avoid was someone trying to handle the `call` as it it as not an upgrade which would be wrong. I see your point on middleware but since current middleware doesn't know about WebSockets I wonder if another approach might be to `call` from the `on_open` or what ever the alternative is.

Reply:
> What I was trying to avoid was someone trying to handle the call as it it as not an upgrade which would be wrong. 

I think this issue would be easily solved by the client getting an unexpected response and failing to connect (404 error, or 200 OK instead of a 400 error or 101 upgrade).

There's a lot of added complexity that a separate `on_open(env)` creates. This might hinder adaptation by major framework authors that rely heavily on middleware (such as Rails / Sinatra).

The application object itself, as well as any middleware (logging, authentication) - these all need to be readjusted just to change the design from an `if` statement to a separate function call.

I like the cleanliness of your idea, but I really doubt we could get users behind us on that one.

Reply:
Help me understand how the middleware is used to reject a WebSockets connection. There does not seem to be an option in the Rack spec to raise from a `call`. How would middleware stop a connection?

Reply:
Middleware looks something like this:

```ruby
class PassThroughMiddleware
   def initialize app
      @app = app
   end
   def call env
      @app.call(env)
   end
end
```

This creates an endless chain, where `app` isn't really the application, it's just the next middleware in the chain. No middleware really knows which `app` is the actual app.

This allows a "filtering" middleware to return a result without invoking the app.

```ruby
class YesNoMiddleware
   def initialize app
      @app = app
      @counter = 1
   end
   def call env
      @counter += 1
       return [400, {}, ["Try Again"]] if @counter.even?
      @app.call(env)
   end
end
```

The rack specification allows the `use` DSL command to stack the middleware and `run` to build all the middleware and app into a nested structure... it's something similar (but more complex to this:

$middleware = []

def use klass
   $middleware << klass
end

def run app
   while($middleware.any?)
     app = $middleware.pop.new(app)
   end
   # start server with the new, nested, app object
end